kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -648,4 +651,47 @@ protected static void validateSupportedType(DataType dataType) {
       throw unsupportedDataType(dataType);
     }
   }
+
+  /**
+   * If IcebergWriterCompatV1 is enabled, we need to ensure that map struct keys don't change. This
+   * validates that
+   */
+  private static void validateNoMapStructKeyChanges(
```

Comment:
I wonder if this check should be nested deeper within the schema evolution checks? Then we can use the `SchemaChanges` class @amogh-jahagirdar added to just check the "updatedColumns"

But I'm also okay with this if it's not any easier/better. Maybe it's fewer schema traversals however? Since I think this method alone involves at least 2

Reply:
Actually that's much nicer, thanks for the suggestion I didn't see the SchemaChanges stuff

Reply:
I was actually picturing nesting this even further in the existing code, I think it can just be an extra check in `validateFieldCompatibility`. There we already have a branch where we have established a field update where we go from map type -> map type (skips a lot of this code here) and then all we need to add there is the additional check if `icebergWriterCompatV1` is enabled & keyType is struct

Reply:
I think this should be super concise and simple there just something like
```
    } else if (existingField.getDataType() instanceof MapType
        && newField.getDataType() instanceof MapType) {
      MapType existingMapType = (MapType) existingField.getDataType();
      MapType newMapType = (MapType) newField.getDataType();

      if (icebergWriterCompatV1Enabled && existingMapType.getKeyType() instanceof StructType && newMapType.getKeyType() instanceof StructType) {
               // Require existingMapType.getKeyType() equals newMapType.getKeyType()
      }

      validateFieldCompatibility(existingMapType.getKeyField(), newMapType.getKeyField());
      validateFieldCompatibility(existingMapType.getValueField(), newMapType.getValueField());
    }
```

Reply:
Ohh man, can't believe I missed that those checks already exist. Thanks! That's obviously much better.